*: Consistent Error/Warning handling for SeriesSet iterator: Allowing Async Select

kakkoyun opened the pull request:
Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

As discussed in https://github.com/prometheus/prometheus/pull/7198, this PR attempts to fix https://github.com/prometheus/prometheus/issues/6878, using deferred SeriesSet iterator.

Currently, Prometheus executes PromQL “.Select” queries sequentially. Particularly for the remote storages, this implementation wastes CPU cycles while waiting for independent queries.

Requirements that gathered from the comments of https://github.com/prometheus/prometheus/issues/6878 and https://github.com/prometheus/prometheus/pull/6887.

#### Requirements
* Should NOT use all the available CPU.
  * Limit concurrency
* PromQL should not care about how the querier is implemented, it should always be the same logic.
* Should avoid introducing changes to LOCAL storage.
* Should avoid spawning goroutines if there is just a single selector.

#### Changes
This PR fulfils the requirements listed above by introducing changes to `storage.SeriesSet`, `storage.Querier`, `promql/engine` and `storage/remote`.
* Add `Warnings()` to `storage.SeriesSet` https://github.com/prometheus/prometheus/commit/18dddfb771f0a8711e9158dc624bd8d18beaf544
* Only return `storage.SeriesSet` from `storage.Querier.Select()` https://github.com/prometheus/prometheus/commit/354a74630056f5b58251376eb96888cc22f1d1c5
* Modify `promql/engine` to propogate errors at evaluation stage. https://github.com/prometheus/prometheus/commit/72203b0f7b29bb83fa5716ae026cd576428039da
* And adds async implementation of SeriesSet to remote reads https://github.com/prometheus/prometheus/commit/1c28c9fb31ee4dee5c4a54f204436c9d2c1e815e

This PR could be separated into smaller PRs if needed.

#### TODO
* [ ] Benchmarks

brian-brazil commented on promql/engine.go:
```diff
@@ -997,29 +1005,32 @@ func (ev *evaluator) rangeEval(f func([]parser.Value, *EvalNodeHelper) Vector, e
-		VectorSelector: vs,
-	}
-	for _, s := range val {
+func (ev *evaluator) evalSubquery(subq *parser.SubqueryExpr) (*parser.MatrixSelector, storage.Warnings) {
```
In a big PR like this, it's best not to make incidental style changes like in this function. It makes the diff bigger than it needs to be.

bwplotka replied:
Yea agree, especially that  the logic is different: Series: make([]storage.Series, 0, len(mat)),

brian-brazil commented on promql/value.go:
```diff
@@ -186,12 +186,16 @@ func (m Matrix) ContainsSameLabelset() bool {
 	return false
 }
 
+// Warnings holds evaluation warnings. Currently, it is just a type alias to
```
This is not what type aliasing should be used for, using storage.Warnings is fine as we use storage in other parts of the API too.

bwplotka commented on promql/engine.go:
```diff
@@ -648,19 +644,13 @@ func (ng *Engine) findMinTime(s *parser.EvalStmt) time.Time {
-		warnings  storage.Warnings
-		err       error
-	)
+func (ng *Engine) populateSeries(ctx context.Context, querier storage.Querier, s *parser.EvalStmt) {
```
hm? :thinking: 

```suggestion
func (ng *Engine) asyncPopulateSeries(ctx context.Context, querier storage.Querier, s *parser.EvalStmt) {
```

kakkoyun replied:
It doesn't necessarily need to be async. It depends on implementation. It puts an unexpanded SeriesSet to the statement node. I can still come up with a better name.

bwplotka replied:
Cool :+1: It's fine. Not a big deal, populate makes sense, just it does not mean things are populated when function finishes.

bwplotka commented on promql/engine.go:
```diff
@@ -742,34 +724,48 @@ func extractGroupsFromPath(p []parser.Node) (bool, []string) {
+	return res, ws, it.Err()
 }
 
+type errorWithWarnings struct {
```
What is this? I don't  think we need to preserve warnings on error? cc @brian-brazil what would be your opinion? :thinking:

brian-brazil replied:
We don't have to, but it wouldn't hurt.

bwplotka replied:
Cool, maybe it would be good candidate to do in following PR

bwplotka commented on promql/engine.go:
```diff
@@ -880,17 +883,20 @@ func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.L
 // step.  The return value is the combination into time series of all the
 // function call results.
-func (ev *evaluator) rangeEval(f func([]parser.Value, *EvalNodeHelper) Vector, exprs ...parser.Expr) Matrix {
+func (ev *evaluator) rangeEval(f func([]parser.Value, *EvalNodeHelper) (Vector, storage.Warnings), exprs ...parser.Expr) (Matrix, storage.Warnings) {
```
I would expect error as well here :thi

kakkoyun replied:
Error handling is a bit different in engine code. It panics when it encounters error and recovers in a top-level `Eval` method, hence we don't return errors in here and similar helper functions.

bwplotka commented:
But overall let's try to minimize changes (:

kakkoyun commented:
@bwplotka @brian-brazil I have removed remote read changes will send another PR for those.
Should I split this one as well? I can split it into two: One for SeriesSet changes and one for the Querier and Engine changes. What do you think?

bwplotka commented:
I think it's fine together. Buildable and complete

liguozhong commented:
We really need this feature
![image](https://user-images.githubusercontent.com/9583245/82134701-ed926a00-982c-11ea-92df-0e513c41b19c.png)

kakkoyun commented:
@liguozhong There will a subsequent PR to actually implement async remote read. Those pieces removed, not to make this PR any larger.

kakkoyun commented:
@bwplotka @krasi-georgiev @brian-brazil Would you take another look at this?

bwplotka commented on promql/engine.go:
```diff
@@ -742,34 +724,48 @@ func extractGroupsFromPath(p []parser.Node) (bool, []string) {
 	switch e := expr.(type) {
 	case *parser.MatrixSelector:
-		checkForSeriesSetExpansion(ctx, e.VectorSelector)
+		if ws := checkAndExpandSeriesSet(ctx, e.VectorSelector); len(ws) > 0 {
```
It's not an error, so we can just return ws and without check return below.

If we do like this it looks like warnings should be returned in certain case which is not true and might mislead here (:

bwplotka commented on promql/engine.go:
```diff
@@ -742,34 +724,48 @@ func extractGroupsFromPath(p []parser.Node) (bool, []string) {
 			} else {
 				e.Series = series
 			}
+			if len(ws) > 0 {
```
No need for this, let's put to function level `ws` and return below

bwplotka commented on promql/engine.go:
```diff
@@ -742,34 +724,48 @@ func extractGroupsFromPath(p []parser.Node) (bool, []string) {
 		default:
 		}
 		res = append(res, it.At())
+		ws = append(ws, it.Warnings()...)
```
Curious, why `.Warnings` is not per SeriesSet? Like `Err`? I think that would enough and avoid some overhead if there are 1 millions series... WDYT @brian-brazil ? 

This would be similar to the existing flow then.

kakkoyun replied:
It's per series set. This a mistake. I'll fix it.

bwplotka replied:
Got offline  info, it's bug only, so all good after we fix this

brian-brazil commented on web/api/v1/api.go:
```diff
@@ -1207,10 +1204,7 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
-				}
+				set := querier.Select(true, hints, filteredMatchers...)
 
 				return remote.StreamChunkedReadResponses(
```
This doesn't seem to be handling warnings.

kakkoyun replied:
Previously, warnings were being ignored. How should we proceed with it in here? Should we log warnings? Should we include the warnings in the response?

brian-brazil replied:
They shouldn't be even happening currently and remote read doesn't support them, so logging is probably fine.

bwplotka replied:
Yea, we talked about this in chunk iterator "in progress" PRs, I think just TODO is fine for now

bwplotka replied:
I think back then we said it will be too spammy  :thinking:

brian-brazil commented on web/api/v1/api.go:
```diff
@@ -1239,10 +1233,7 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
-				if err != nil {
-					return err
-				}
+				set := querier.Select(false, hints, filteredMatchers...)
```
Ditto

brian-brazil commented on storage/fanout.go:
```diff
@@ -471,26 +453,45 @@ func newGenericMergeSeriesSet(sets []genericSeriesSet, querier *mergeGenericQuer
 
 func (c *genericMergeSeriesSet) Next() bool {
+	if c.heap == nil {
+		// Sets need to be pre-advanced, so we can introspect the label of the
```
This is only advancing this select, we need to advance all the selects done against the querier to spot if any of them failed so that we can filter out responses from remotes that fail some of their requests.

kakkoyun replied:
Could you please elaborate this one a bit more?

bwplotka replied:
I think what Brian is saying you exactly did below @kakkoyun , so we make sure `Next` is done so we ensure no error is there.

However TBH we  can simplify this whole thing by one simple trick: We can make sure remote read Select will behave like this (if select fails we only return warning not error). This will make the flow super easy (: We can name it even `nonErrReadQuerier` or something.

brian-brazil replied:
You're still only advancing the sets of this Select, not the sets of all Selects.

liguozhong commented:
> @liguozhong There will a subsequent PR to actually implement async remote read. Those pieces removed, not to make this PR any larger.

Thanks, looking forward to seeing this feature in the 2.19.0 release.

bwplotka commented on promql/engine.go:
```diff
@@ -1203,20 +1215,23 @@ func (ev *evaluator) eval(expr parser.Expr) parser.Value {
 
 	case *parser.UnaryExpr:
-		mat := ev.eval(e.Expr).(Matrix)
+		var (
```
Sometimes we do `var(` sometimes direct `:=` , let's pick one (:

bwplotka commented on storage/fanout.go:
```diff
@@ -257,53 +257,32 @@ func NewMergeChunkQuerier(primaryQuerier ChunkQuerier, queriers []ChunkQuerier,
-		if qryResult.wrn != nil {
-			warnings = append(warnings, qryResult.wrn...)
-		}
-		if qryResult.selectError != nil {
```
We might want to ensure correct logic of this in the remote read Select implementation (:

bwplotka replied:
Is this addressed?

kakkoyun replied:
Errors are returned by remote read are now wrapped with `ErrSeriesSet`.

```go
// Select implements storage.Querier and uses the given matchers to read series sets from the Client.
func (q *querier) Select(sortSeries bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
	query, err := ToQuery(q.mint, q.maxt, matchers, hints)
	if err != nil {
		return storage.ErrSeriesSet(err)
	}

	remoteReadGauge := remoteReadQueries.WithLabelValues(q.client.remoteName, q.client.url.String())
	remoteReadGauge.Inc()
	defer remoteReadGauge.Dec()

	res, err := q.client.Read(q.ctx, query)
	if err != nil {
		return storage.ErrSeriesSet(fmt.Errorf("remote_read: %v", err))
	}

	return FromQueryResult(sortSeries, res)
}
```

Is this what you meant?

bwplotka commented on storage/fanout.go:
```diff
@@ -502,6 +506,9 @@ func (c *genericMergeSeriesSet) Next() bool {
 			if set.Next() {
 				heap.Push(&c.heap, set)
 			}
+			if err := set.Err(); err != nil && c.err == nil {
```
I think we can skip those, and just do return false in case of error and then allow for loop to loop through `set.Err` as we had before, just easier ..

bwplotka replied:
No a blocker here, up to you

bwplotka commented on storage/interface.go:
```diff
@@ -154,6 +154,8 @@ type SeriesSet interface {
 	Next() bool
 	At() Series
 	Err() error
+	// A collection of warnings for the whole set.
```
Let's to the same for Err. 

Also let's mention that it also is there for both error and non error sets

bwplotka commented on storage/remote/read.go:
```diff
@@ -116,13 +116,10 @@ type externalLabelsQuerier struct {
-		return nil, warnings, err
-	}
-	return newSeriesSetFilter(s, added), warnings, nil
+	s := q.Querier.Select(sortSeries, hints, m...)
```
Let's inline I think (:

bwplotka commented on tsdb/querier.go:
```diff
@@ -125,31 +119,22 @@ type verticalQuerier struct {
-		return nil, ws, err
-	}
-	return newMergedVerticalSeriesSet(a, b), ws, nil
+	a := q.sel(sortSeries, hints, qs[:l], ms)
```
inline please if we can as per https://thanos.io/contributing/coding-style-guide.md/#avoid-defining-variables-used-only-once :hugs:

bwplotka commented on tsdb/querier.go:
```diff
@@ -509,9 +499,11 @@ func (s *mergedSeriesSet) nextAll() {
 	for _, ss := range s.all {
 		if ss.Next() {
 			s.buf = append(s.buf, ss)
+			s.ws = append(s.ws, ss.Warnings()...)
```
:+1:

bwplotka commented on tsdb/querier.go:
```diff
@@ -509,9 +499,11 @@ func (s *mergedSeriesSet) nextAll() {
 		if ss.Next() {
 			s.buf = append(s.buf, ss)
+			s.ws = append(s.ws, ss.Warnings()...)
 		} else if ss.Err() != nil {
```
I would vote for no else but `continue` instead (: https://thanos.io/contributing/coding-style-guide.md/#control-structure-prefer-early-returns-and-avoid-else

bwplotka commented on tsdb/querier.go:
```diff
@@ -132,9 +132,10 @@ func (q *verticalQuerier) sel(sortSeries bool, hints *storage.SelectHints, qs []
-	a := q.sel(sortSeries, hints, qs[:l], ms)
-	b := q.sel(sortSeries, hints, qs[l:], ms)
-	return newMergedVerticalSeriesSet(a, b)
+	return newMergedVerticalSeriesSet(
```
:+1:

bwplotka commented on storage/fanout.go:
```diff
@@ -309,11 +318,27 @@ func (q *mergeGenericQuerier) LabelValues(name string) ([]string, Warnings, erro
+	q.failedQueriers[q.setQuerierMap[set]] = struct{}{}
+}
+
+func (q *mergeGenericQuerier) DoesBelongToPrimary(set genericSeriesSet) bool {
```
does those have to be exported? :thinking:

bwplotka replied:
and if yes, let's comment

brian-brazil replied:
Same question from me, these all appear to be internal stuff

kakkoyun replied:
It doesn't have to be. I've just followed the same fashion with `IsFailedSet`. It's also not used outside `fanout`. Maybe be better to un-export `IsFailedSet`?

kakkoyun replied:
I've removed and inlined `IsFailedSet` as well.

bwplotka commented on storage/fanout.go:
```diff
@@ -282,7 +284,14 @@ func (q *mergeGenericQuerier) Select(sortSeries bool, hints *SelectHints, matche
+		mergeFunc: q.mergeFunc,
+		sets:      seriesSets,
+	}
+	q.mergeSeriesSets = append(q.mergeSeriesSets, set)
```
Can you explain here what is happening here? I don't follow this code.. why have to remember the set during `Select` in `mergeSeriesSets`. This sounds wrong as Select can be done multiple times.

kakkoyun replied:
As @brian-brazil pointed out, we need to make sure all sets from all selects are succeeded https://github.com/prometheus/prometheus/pull/7251#discussion_r426546061 before we merge sets, otherwise we ignore to sets that come from that querier. Thus I added this to remember sets.

Also I've added a comment for this.

brian-brazil commented on storage/fanout.go:
```diff
@@ -429,7 +433,12 @@ type genericMergeSeriesSet struct {
 	currentSets []genericSeriesSet
-	querier     *mergeGenericQuerier
+
+	// 'querier' is optional and can be nil. Pass Querier if you want to retry query in case of failing series set.
```
We never retry, why is this here?

kakkoyun replied:
This is a moved comment from a factor function to the struct. I believe it's an outdated one, I'll remove.

brian-brazil commented on storage/fanout.go:
```diff
@@ -457,40 +466,67 @@ func NewMergeChunkSeriesSet(sets []ChunkSeriesSet, merger VerticalChunkSeriesMer
 }
 
 func (c *genericMergeSeriesSet) Next() bool {
+	if c.querier != nil { // since querier is optional.
```
Since

brian-brazil commented on storage/fanout.go:
```diff
@@ -330,11 +318,27 @@ func (q *mergeGenericQuerier) LabelValues(name string) ([]string, Warnings, erro
 	return isFailedQuerier
 }
 
+func (q *mergeGenericQuerier) AdvanceAllSets() {
```
This is only used once, you can inline this.

brian-brazil commented on storage/fanout.go:
```diff
@@ -330,11 +318,27 @@ func (q *mergeGenericQuerier) LabelValues(name string) ([]string, Warnings, erro
 	return mergeStringSlices(results), warnings, nil
 }
 
+func (q *mergeGenericQuerier) MarkAsFailedSet(set genericSeriesSet) {
```
This is also only used once

kakkoyun commented:
> Sorry, but I don't follow last commit ): It might need some explanation or redesign. What do you want to achieve? What was wrong without this commit?

I have added the last two commits to address the issue that has raised by Brian. https://github.com/prometheus/prometheus/pull/7251#discussion_r426546061

I'm happy to come up with a better design if it doesn't match the requirements. And also I've already addressed the issue with exported methods by removing and inlining them.

brian-brazil commented on storage/fanout.go:
```diff
@@ -539,14 +567,28 @@ func (c *genericMergeSeriesSet) At() Labels {
 
+func (c *genericMergeSeriesSet) Warnings() Warnings {
+	for _, set := range c.sets {
+		if ws := set.Warnings(); len(ws) > 0 {
```
Hmm, do we need the if here? I'd expect append to deal with the empty variadic.

bwplotka commented on promql/engine.go:
```diff
@@ -742,34 +724,43 @@ func extractGroupsFromPath(p []parser.Node) (bool, []string) {
+	return res, it.Warnings(), it.Err()
 }
 
+type errorWithWarnings struct {
```
```suggestion
type errWithWarnings struct {
```

Tiny nit

bwplotka commented on rules/manager.go:
```diff
@@ -727,6 +722,13 @@ func (g *Group) RestoreForState(ts time.Time) {
 
+			if err := sset.Err(); err != nil {
+				// Querier Warnings are ignored. We do not care unless we have an error.
+				level.Error(g.logger).Log("msg", "Failed to restore 'for' state",
```
Incorrent formatting ):

bwplotka commented on storage/fanout.go:
```diff
@@ -457,40 +446,76 @@ func NewMergeChunkSeriesSet(sets []ChunkSeriesSet, merger VerticalChunkSeriesMer
+}
+
+func (c *genericMergeSeriesSet) advance() {
+	if !c.advanced {
```
This has unnecessary indent.. (: Can you just return if `c.advanced?`

bwplotka replied:
This is a code smell as well. Why would you `advance` if it was advanced before? It's better if caller will make sure this will not happen, otherwise we can end up in unnecessary  invoking `advance` thousand times `just in case`. This creates confusion and is extremely hard to follow. I would fix this TBH before merging.

bwplotka commented on storage/fanout.go:
```diff
@@ -457,40 +446,76 @@ func NewMergeChunkSeriesSet(sets []ChunkSeriesSet, merger VerticalChunkSeriesMer
+			}
+
+			if err := set.Err(); err != nil {
+				if c.querier != nil { // Since querier is optional.
```
Wait, so if querier is nil, then we ignore error? That sounds bad.

kakkoyun commented:
@bwplotka That'd awesome, let's review it closer together.

kakkoyun commented on storage/fanout.go:
```diff
@@ -512,8 +537,11 @@ func (c *genericMergeSeriesSet) Next() bool {
 		c.currentLabels = c.heap[0].At().Labels()
 		for len(c.heap) > 0 && labels.Equal(c.currentLabels, c.heap[0].At().Labels()) {
 			set := heap.Pop(&c.heap).(genericSeriesSet)
-			if c.querier != nil && c.querier.IsFailedSet(set) {
```
@bwplotka To address your comment in here https://github.com/prometheus/prometheus/pull/7251#discussion_r430309497, querier was optional previously.

bwplotka commented on storage/fanout.go:
```diff
@@ -457,40 +446,76 @@ func NewMergeChunkSeriesSet(sets []ChunkSeriesSet, merger VerticalChunkSeriesMer
 }
 
 func (c *genericMergeSeriesSet) Next() bool {
+	if c.querier != nil { // Since querier is optional.
```
Why we check `c.querier` twice?

bwplotka commented:
As promised: https://github.com/kakkoyun/prometheus/pull/3

kakkoyun commented:
Just adding the last comment that I had in https://github.com/kakkoyun/prometheus/pull/3 in case we forget.

> Looks great much easier on the eyes :)
> I had a quick look at it and I want to mediate on it a bit more.
> 
> Besides the nits, my major concern is; this removes the ability to exec async request against queriers.
> Both mergeSeriesSet and secondarySeriesSet constructed with newGenericMergeSeriesSet() and it advances all the set at least once. So fanned out .Select calls are must be completed before leaving the overarching .Select call. Thus I had put the deferred set wrapper in place.
> 
> What do you think?

bwplotka commented:
Ok I think we got consensus offline with @kakkoyun 

The idea for implementation of async/lazy series set is to schedule request in `Select` and do blocking receive on first `Next` With current code we can do exactly that :hugs: 

This should work well. Still, I am adding some last minutes tests for error handling check, but overall it should be good for review @brian-brazil :hugs:

bwplotka commented:
> Where has the logic gone that if a Select for a secondary fails, then all other Selects for that secondary must not return anything?

Ah that's why this was needed .... =D Now all clear will add back with proper comment and tests against this requirement :+1: 

Thanks for review @brian-brazil 

(I am on it as @kakkoyun is OOO today)

bwplotka commented:
Tried to summarise our confusion part here:

## Current Flow
![Prometheus PromQL v2 19 Flow](https://user-images.githubusercontent.com/6950331/83546026-f99e4d00-a4f7-11ea-939a-9984a16d7e47.png)

## Partial Result Issue
![Prometheus PromQL_ Partial Response issue](https://user-images.githubusercontent.com/6950331/83546017-f60ac600-a4f7-11ea-9f17-4a0ede448ddb.png)

bwplotka commented on storage/remote/storage.go:
```diff
@@ -145,6 +145,7 @@ func (s *Storage) Querier(ctx context.Context, mint, maxt int64) (storage.Querie
 		}
 		queriers = append(queriers, q)
 	}
+	// All of those queriers are considered best effort, so we pass them as secondary queriers.
```
Why? This is true, all of them are secondary (like it was before)

brian-brazil replied:
This comment has nothing to do with the code that's here.

bwplotka commented:
All should be addressed. I only need to add a few more test use cases, will add tomorrow. Otherwise PTAL @kakkoyun @brian-brazil  (:

kakkoyun commented on storage/fanout.go:
```diff
@@ -188,153 +187,108 @@ func (f *fanoutAppender) Rollback() (err error) {
+		seriesSets = append(seriesSets, r)
 	}
-	return newGenericMergeSeriesSet(seriesSets, q, q.mergeFunc), warnings, nil
+	return newGenericMergeSeriesSet(seriesSets, q.mergeFn)
```
@bwplotka Even though `Select` from `SecondQuerier` returns a lazy set, in `newGenericMergeSeriesSet` we call `Next()` to advance sets so we unwrap lazy set before we leave this function. I don't think this is what we want, is it?

brian-brazil commented on storage/secondary.go:
```diff
@@ -13,25 +13,39 @@
+// All other SeriesSet will be return no response as well. This ensures consistent partial response strategy, where you
+// have either full results or none from each secondary Querier.
+// NOTE: This works well only for implementations that only fail during first Next() (e.g fetch from network). If implementation fails
+// during further iterations, set will panic. If Select is invoked after first Next of any returned SeriesSet, querier will panic.
```
Select will panic?

bwplotka replied:
Yes, what else to do if Select will be run after `once` will already complete?

brian-brazil replied:
My point was s/set/Select/

brian-brazil commented on storage/secondary.go:
```diff
@@ -50,18 +64,88 @@ func (s *secondaryQuerier) LabelNames() ([]string, Warnings, error) {
+			// errors. This is to ensure we support consistent partial failures.
+			var errs tsdb_errors.MultiError
+			for i, set := range s.asyncSets {
+				if set.Next() {
```
Hmm, might this lose us the first series?

bwplotka replied:
Yea, not as we lazySeriesSet cares for that

bwplotka commented on storage/fanout_test.go:
```diff
@@ -596,3 +639,261 @@ func BenchmarkMergeSeriesSet(b *testing.B) {
+			expectedLabels: []string{"b", "c"},
+		},
+		{
+			// TODO(kakkoyun): ?
```
looks like some left over comment? The case looks good now.

kakkoyun commented on rules/manager.go:
```diff
@@ -727,6 +722,17 @@ func (g *Group) RestoreForState(ts time.Time) {
 				}
 			}
 
+			if err := sset.Err(); err != nil {
```
Yes, apparently we haven't.

brian-brazil commented on storage/fanout.go:
```diff
@@ -37,8 +38,15 @@ type fanout struct {
 // through to multiple underlying storages.
+//
+// The difference between primary and secondary Storage is only for read (Querier) path and it goes as follows:
+// * If the primary querier returns an error, the any of the Querier operation will fail.
```
then any of the Querier's operations

brian-brazil commented on storage/fanout.go:
```diff
@@ -37,8 +38,15 @@ type fanout struct {
+//
+// The difference between primary and secondary Storage is only for read (Querier) path and it goes as follows:
+// * If the primary querier returns an error, the any of the Querier operation will fail.
+// * If the any secondary querier returns an error the result is discarded, fanout returns only warnings but the Querier
```
If any secondary querier returns an error the result from that queries is discarded. The overall operation will succeed, and the error from the secondary querier will be returned as a warning.

brian-brazil commented on storage/secondary.go:
```diff
@@ -0,0 +1,107 @@
+			if errs.Err() != nil {
+				// One failed, ensure all sets returns nothing. (All or nothing logic).
+				for i, set := range s.asyncSets {
+					s.asyncSets[i] = warningsOnlySeriesSet(set.Warnings())
```
Hmm, would this mean the warnings will be duplicated many times to the user? Might we want to do a noop series set for the others instead?

bwplotka replied:
This is very valid point, let's do that and ensure on our new tests. I remember I wanted to do it , but forgot.
